### PR TITLE
feat(cli): list subcommand

### PR DESCRIPTION
**Pull request by j03-dev**
-

-

#### Migration notes

...
## Add `list` subcommand to List typegraph(s) from typegate

- [ ] The change comes with new or modified tests
- [ ] Hard-to-understand functions have explanatory comments
- [ ] End-user documentation is updated to reflect the change
- [x] Change Something on CONTRIBUTING.md `fix contributing`

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -133,7 +133,7 @@ ghjk x dev-compose all # or only the envs required (e.g. base prisma s3)
 ghjk x test-e2e # all tests
```
Comment (Yohe-Am) [inline CONTRIBUTING.md:133]: Care to add that missing `build-graph` step here?

File: meta-cli/src/cli/list.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+impl List {
+    async fn get_more_info(&self) -> Result<Vec<TypegraphObject>> {
+        let client = reqwest::Client::new();
+
+        let mut headers = HeaderMap::new();
+        headers.insert(
+            AUTHORIZATION,
+            HeaderValue::from_str(&format!("Basic {}", "YWRtaW46cGFzc3dvcmQ="))?, // I don't know if it right, may be get from env
```
Comment (Yohe-Am) [inline meta-cli/src/cli/list.rs:90]: thought: you can get this info from the config. You just `base64(`${user}:${password}`). Or better yet, use the `post` method on the `Node` type found in the `common` crate. Grep for other usages in the cli for examples.

**Review by Yohe-Am — APPROVED** [review]

File: CONTRIBUTING.md (hunk 1 above)
Comment (j03-dev) [reply CONTRIBUTING.md:134]: ok, done!

File: meta-cli/src/cli/list.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            query {
+                typegraphs {
+                    name
+                    url
+                }
+            }
+        "#;
+
+        let response = client
+            .post("http://localhost:7891/typegate") // this too, may be get from env
```
Comment (destifo) [inline meta-cli/src/cli/list.rs:103]: I think you can also get this from the [config](https://github.com/metatypedev/metatype/blob/1d62c0320fb275ce9efabf772de2d604780105d7/examples/metatype.yaml#L3C5-L3C33),

**Review by destifo — COMMENTED** [review]
great work all around 🙌

File: meta-cli/src/cli/list.rs (hunk 3 above)
Comment (j03-dev) [reply meta-cli/src/cli/list.rs:103]:
```rust
impl Default for NodeConfig {
    fn default() -> Self {
        NodeConfig {
            url: "http://localhost:7890".parse().unwrap(),
            prefix: None,
            username: None,
            password: None,
            env: HashMap::default(),
            secrets: HashMap::default(),
        }
    }
}
```

the url is not same, the graphql is on 7891,

File: meta-cli/src/cli/list.rs (hunk 2 above)
Comment (j03-dev) [reply meta-cli/src/cli/list.rs:90]: I already change it, but I wait @Natoandro to check why path in Typegraph is None


```rust
    async fn get_more_info(&self, dir: PathBuf) -> Result<Vec<TypegraphObject>> {
        let query = r#"
            query {
                typegraphs {
                    name
                    url
                }
            }
        "#;
        let mut node_config = NodeConfig::default();
        node_config.url = Url::parse("http://localhost:7891")?;
        let node = node_config.build(dir).await?;
        let response = node
            .post("/typegate")
            .unwrap()
            .gql(query.into(), None)
            .await?;

        response
            .data("typegraphs")
            .map_err(|err| anyhow::anyhow!(err))
    }

```

File: meta-cli/src/cli/list.rs (hunk 3 above)
Comment (afmika) [reply meta-cli/src/cli/list.rs:103]: I would suggest you try change your local config to point at `localhost:7891` for testing purposes instead, the more generic we are the better it is. Or maybe use something similar to [this](https://github.com/metatypedev/metatype/blob/main/typegate/tests/e2e/templates/templates_test.ts#L50) for quick testing.

File: dev/tasks-install.ts [hunk 4]
```diff
@@ -9,13 +9,18 @@ const tasks: Record<string, DenoTaskDefArgs> = {
...
-      if (/(Ubuntu|Debian)/.test(uname)) {
+      const osRelease = await $`cat /etc/os-release`.text();
+      if (/(Ubuntu|Debian)/.test(osRelease)) {
         console.log(
           `sudo apt update && ` +
             `sudo apt install -y --no-install-recommends ` +
             `gcc-multilib pkg-config libssl-dev libclang-dev perl make`,
         );
+      } else if (/Fedora|Red Hat|CentOS/.test(osRelease)) {
+        console.log(
```
Comment (zifeo) [inline dev/tasks-install.ts:20]: 👍

File: meta-cli/src/cli/list.rs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    #[tracing::instrument]
+    async fn run(&self, args: ConfigArgs) -> Result<()> {
+        let dir = args.dir()?;
+        let config_path = args.config.clone();
+
+        let config = Arc::new(Config::load_or_find(config_path, &dir)?);
+
+        let console = ConsoleActor::new(Arc::clone(&config)).start();
+
+        let action_generator = SerializeActionGenerator::new(
```
Comment (zifeo) [inline meta-cli/src/cli/list.rs:37]: We will likely have other similar "simple command", do you see how we could simplify that setup into some utils to keep the run DRY?

File: meta-cli/src/cli/list.rs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        format!(
+            "{:<50} {:<20} {:<30}",
+            self.name,
+            self.path.as_deref().unwrap_or("None"),
+            self.url.as_deref().unwrap_or("None")
+        )
+    }
+
+    fn print_table_header() {
+        println!("{:<50} {:<20} {:<30}", "NAME", "PATH", "URL");
```
Comment (zifeo) [inline meta-cli/src/cli/list.rs:173]: Some extensions (as this is your first exploratory PR, later tasks shall be fully described from the start) that may be worth implementing :
- add a column to specific which under which target the typography is deployed
- the `local` target could refer to the dev instance
- the target could be `-` for a non deployed tg

**Review by zifeo — COMMENTED** [review]
Nice first contrib! Do not forget to sign the CLA and rebase your code :)

File: meta-cli/src/cli/list.rs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+            let mut table = Table::new(tg.name().unwrap());
+            table.set_path(tg.get_path().ok());
+            tables.push(table);
+        }
+
+        if let Ok(more_info) = self.get_more_info(dir).await {
+            for tg in more_info {
+                let mut table = Table::new(tg.name);
+                table.set_url(Some(tg.url));
+                tables.push(table);
```
Comment (Natoandro) [inline meta-cli/src/cli/list.rs:126]: /Merge entries with the same typegraph name?

File: meta-cli/src/cli/list.rs [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        for table in tables {
+            println!("{}", table.to_table_row());
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Table {
```
Comment (Natoandro) [inline meta-cli/src/cli/list.rs:140]: I think you meant `TableRow` or `TableEntry` here.

File: typegraph/node/sdk/src/tg_manage.ts [hunk 9]
```diff
@@ -26,6 +26,9 @@ export class Manager {
       case "deploy":
         await this.#deploy();
         break;
+      case "list":
+        await this.#list();
```
Comment (Natoandro) [inline typegraph/node/sdk/src/tg_manage.ts:30]: This won't be called as you use the `SerializeActionGenerator`...

**Review by Natoandro — CHANGES_REQUESTED** [review]
Great !

File: meta-cli/src/cli/list.rs (hunk 3 above)
Comment (Natoandro) [reply meta-cli/src/cli/list.rs]: I think we need to add the same `--target` flag as in the serialize subcommand.

File: typegraph/node/sdk/src/tg_manage.ts (hunk 9 above)
Comment (j03-dev) [reply typegraph/node/sdk/src/tg_manage.ts:30]: ah, ok thanks

File: meta-cli/src/cli/list.rs (hunk 7 above)
Comment (j03-dev) [reply meta-cli/src/cli/list.rs:126]: No, I just create table to display them

**Pull request closed without merging**